### PR TITLE
feat: add SM90 linear logp backward fast paths

### PR DESCRIPTION
**Pull request by inaniloquentee**
## Summary

This PR adds SM90/Hopper fast paths for `linear_logp`, focused on making the training backward path faster instead of only optimizing forward.

Before this change, the SM90 path optimized forward by avoiding materialization of full `[N, V]` logits, but backward still used the portable chunked recompute path. This PR adds fused and optional fast backward paths while keeping fallback behavior available.

## What Changed

- Added SM90 fused `linear_logp` backward support.
- Added tensor-parallel local-shard support for global target ids.
- Added bf16 save-logits / fused-tile full-gradient fast paths.
- Added optional save-probs output-only paths.
- Added optional streaming/tiled backward paths.
- Added dispatch logic to skip unused gradients.
- Added tests for SM90 backward, optional fast paths, and TP dispatch selection.

## Optional Fast Paths

The new performance paths are not all forced by default. They are selected only when the hardware, dtype, shape, gradient requirements, and environment settings match.

Key controls:

- `RL_KERNEL_LINEAR_LOGP_FUSED_BACKWARD=0`
  disables fused backward and falls back to the portable path.
- `RL_KERNEL_LINEAR_LOGP_FUSED_BWD_PRECISION=fp32`
  forces the stricter fp32-style backward path.
- `RL_KERNEL_LINEAR_LOGP_SAVE_PROBS_BF16=1`
  enables save-probs bf16 paths.
- `RL_KERNEL_LINEAR_LOGP_STREAMING_BWD=1`
  enables streaming output-only backward.
- `RL_KERNEL_LINEAR_LOGP_FUSED_TILE_BWD_FULL=...`
  enables fused-tile full-gradient paths.

Benefits of these optional paths:

- **save-probs**: avoids recomputing probabilities in backward, useful when trading memory for speed is acceptable.
- **streaming backward**: reduces workspace pressure for output-layer-only training, especially when `hidden` does not require gradients.
- **tiled backward**: avoids processing the full vocab workspace at once and is useful for large-vocab configurations.
- **fused backward**: reduces Python chunk-loop overhead and improves the main training backward path.

Unsupported or disabled cases still fall back to Triton/PyTorch paths.

## Benchmark Setup

External full-pipeline validation was run with vime + RL-Kernel:

- Model: `Qwen3-30B-A3B`
- GPUs: `8x H100 80GB`
- Parallelism: `TP=2, PP=1, CP=1, EP=8`
- Rollout: full vLLM rollout, no debug rollout
- Number of rollouts: `12`
- Baseline: native output layer + native logprob
- Candidate: RL-Kernel CUDA `linear_logp`
- Candidate target path:
  `Using fused-tile bf16 full-gradient tensor-parallel linear_logp fast path.`
- Fallback count: `0` for completed candidate runs

## Operator-Level Results

The primary comparison is the `linear_logp` operator timing, not end-to-end step time.

| Config | Tokens/call baseline | Tokens/call candidate | Fwd CUDA ms baseline | Fwd CUDA ms candidate | Fwd speedup | Fwd+Bwd CUDA ms baseline | Fwd+Bwd CUDA ms candidate | Fwd+Bwd speedup | Fallback |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| T1 | 796.44 | 796.44 | 5.14 | 3.51 | 1.46x | 15.24 | 12.37 | 1.23x | 0 |
| T2 | 1820.44 | 1820.44 | 7.78 | 3.36 | 2.32x | 18.56 | 10.37 | 1.79x | 0 |
| T3 | 6769.78 | 6826.67 | 14.52 | 7.62 | 1.91x | 33.96 | 18.50 | 1.84x | 0 |

Notes:

- `Fwd+Bwd CUDA ms` is the main timing metric.
- Candidate runs hit the intended fused-tile full-gradient TP path.
- Fallback was `0` in completed candidate runs.
- Baseline had occasional native logprob spikes; stable windows were used for the operator comparison.

## Memory

| Config | Peak VRAM GB baseline | Peak VRAM GB candidate |
| --- | ---: | ---: |
| T1 | 20.60 | 19.89 |
| T2 | 24.14 | 22.49 |
| T3 | 49.26 | 46.23 |

The candidate used less peak reserved memory in all completed full-gradient runs, with the largest completed config reducing peak reserved VRAM by about `3.03 GB`.

## Correctness / Stability

Completed 12-rollout runs passed finite-loss checks.

| Config | Baseline abs diff | Candidate abs diff | Loss finite |
| --- | ---: | ---: | --- |
| T1 | 0.02668 | 0.02537 | pass |
| T2 | 0.02264 | 0.02395 | pass |
| T3 | 0.02034 | 0.02224 | pass |

The absolute logprob difference stayed in the same range as baseline.

## Validation

Local validation:

- `pre-commit run --all-files`
- `mypy --ignore-missing-imports rl_engine/`
- `python -m pytest rl_engine/tests/test_dispatch.py -v`
- `PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest tests/test_attention_correctness.py -q -rs`
- `python -m pytest tests/test_attention.py -v -k "not large and not gpu"`
- `python -m pytest tests/test_kv_cache_attention.py -v -k "not large and not gpu"`
- `python -m pytest tests/test_kernel_registry.py -v`
- `mkdocs build --strict -f mkdocs.yaml`

External metrics source:

`vime/vime-RLK-final-metrics-config-matrix.md`

## Risk and Fallback

This is a performance-oriented SM90 change. The main risks are numeric drift from more aggressive bf16/TF32 paths and the extra complexity of multiple specialized kernels.

Mitigations:

- Optional paths are gated.
- Conservative fallback paths remain available.
- Fused backward can be disabled.
- fp32-style validation mode remains available.
- Unsupported hardware/dtypes/shapes fall back to Triton/PyTorch implementations.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
cc @KJLdefeated @EthanZero2Hero PTAL

**Comment by Flink-ddd** [comment]
please resolve the CI error first, This PR supports the work comparing it to Vime, thank you for your work.

**Comment by inaniloquentee** [comment]
> please resolve the CI error first, This PR supports the work comparing it to Vime, thank you for your work.

Fixed pre-commit and CI-pipline

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thank you for work!

**Review by KJLdefeated — APPROVED** [review]
LGTM. Verified with local machine. The core kernels are correct and the speedups are great. Thx.
